src/fsharp/block.fs

```diff
@@ -0,0 +1,222 @@
+module Internal.Utilities.Library.Block
+
+open System.Collections.Immutable
+
+type block<'T> = ImmutableArray<'T>
+type blockbuilder<'T> = ImmutableArray<'T>.Builder
+
+[<RequireQualifiedAccess>]
+module BlockBuilder =
+
+    let create size : blockbuilder<'T> =
+        if size = 0 then
+            ImmutableArray.CreateBuilder()
+        else
+            ImmutableArray.CreateBuilder(size)
+
+[<RequireQualifiedAccess>]
+module Block =
+
+    let empty<'T> = ImmutableArray<'T>.Empty
+
+    let init n (f: int -> 'T) : block<_> =
+        match n with
+        | 0 -> ImmutableArray.Empty
+        | 1 -> ImmutableArray.Create(f 0)
+        | n ->
+            if n < 0 then
+                invalidArg "n" "Below zero."
+
+            let builder = ImmutableArray.CreateBuilder(n)
+            for i = 0 to n - 1 do
+                builder.Add(f i)
+            builder.ToImmutable()
```

Comment:
As long as you keep your builder private and builder.Capacity equals builder.Count (and these are settable) you should really use MoveToImmutable() to avoid another copy. (here, and all the other places)!